Ensure no Watches are running after Watcher is stopped.

jakelandis opened the pull request:
Watcher keeps track of which watches are currently running keyed by watcher name/id.
If a Watch is currently running it will not run the same Watch and will result in a
message : "Watch is already queued in thread pool" and a state: "not_executed_already_queued"

When Watcher is stopped, it will stop watcher (rejecting any new watches), but allow
the currently running watches to run to completion. Waiting for the currently running
Watches to complete is done async to the stopping of Watcher. Meaning that Watcher will
report as fully stopped, but there is still a background thread waiting for all of the
Watches to finish before it removes the Watch from it's list of currently running Watches.

The integration test start and stop watcher between each test. The goal to ensure a clean
state between tests. However, since Watcher can report "yes - I am stopped", but there
are still running Watches, the tests may bleed over into each other, especially on slow
machines.  This can result in errors related to "Watch is already queued in thread pool"
and a state: "not_executed_already_queued", and is VERY difficult to reproduce. This 
may also change the most recent Watcher history document in an unpredictable way. 

This commit changes the waiting for Watches on stop/pause from an aysnc waiting, back to a
sync wait as it worked prior to #30118.  This help ensure that for testing testing scenario
the stop much more predictable, such that after fully stopped, no Watches are running.
This should have little impact if any on production code since Watcher isn't stopped/paused
too often and when it stop/pause it has the same behavior is the same, it will just run on
the calling thread, not a generic thread.

Related: #42409

jakelandis committed:
Ensure no Watches are running after Watcher is stopped.

Watcher keeps track of which watches are currently running keyed by watcher name/id.
If a watch is currently running it will not run the same watch and will result in a
message : "Watch is already queued in thread pool" and a state: "not_executed_already_queued"

When Watcher is stopped, it will stop watcher (rejecting any new watches), but allow
the currently running watches to run to completion. Waiting for the currently running
watches to complete is done async to the stopping of Watcher. Meaning that Watcher will
report as fully stopped, but there is still a background thread waiting for all of the
Watches to finish before it removes the watch from it's list of currently running Watches.

The integration test start and stop watcher between each test. The goal to ensure a clean
state between tests. However, since Watcher can report "yes - I am stopped", but there
are still running Watches, the tests may bleed over into each other, especially on slow
machines.  This can result in errors related to "Watch is already queued in thread pool"
and a state: "not_executed_already_queued", and is VERY difficult to reproduce.

This commit changes the waiting for Watches on stop/pause from an aysnc waiting, back to a
sync wait as it worked prior to #30118.  This help ensure that for testing testing scenario
the stop much more predictable, such that after fully stopped, no Watches are running.
This should have little impact if any on production code since Watcher isn't stopped/paused
too often and when it stop/pause it has the same behavior is the same, it will just run on
the calling thread, not a generic thread.

jakelandis committed:
fix check style

elasticmachine commented:
Pinging @elastic/es-core-features

martijnvg commented on x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/execution/ExecutionService.java:
```diff
@@ -106,7 +105,7 @@
     private final ExecutorService genericExecutor;
 
-    private AtomicReference<CurrentExecutions> currentExecutions = new AtomicReference<>();
+    private CurrentExecutions currentExecutions;
```
Can `CurrentExecutions` remain inside an `AtomicReference`? 
It is read else where without acquiring a lock and as far as I understand it is about making sure that `clearExecutions()` happens in a sync manner, which is possible with keeping the `AtomicReference`?

(also the currentExecutions field can then be made final)

gwbrown replied:
Yes, I think this either needs to be `volatile` or go back to being an `AtomicReference`.

jakelandis replied:
thanks nice catch. I also had messed up the order of "sealing" the concurrent executions.  The change here is now a single line that removes the fork.

martijnvg reviewed (commented):
I left a question, but this looks good otherwise.

jakelandis commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into watcher_stop_less_async

jakelandis committed:
Revert "fix check style"

This reverts commit 9d18274a2635f7b06ea10aea7f317941eca7ed9e.

jakelandis committed:
Revert "Ensure no Watches are running after Watcher is stopped."

This reverts commit 926a671c6f96cc6d786b7da78093f53e75b92f07.

jakelandis committed:
Ensure no Watches are running after Watcher is stopped.

Watcher keeps track of which watches are currently running keyed by watcher name/id.
If a watch is currently running it will not run the same watch and will result in a
message : "Watch is already queued in thread pool" and a state: "not_executed_already_queued"

When Watcher is stopped, it will stop watcher (rejecting any new watches), but allow
the currently running watches to run to completion. Waiting for the currently running
watches to complete is done async to the stopping of Watcher. Meaning that Watcher will
report as fully stopped, but there is still a background thread waiting for all of the
Watches to finish before it removes the watch from it's list of currently running Watches.

The integration test start and stop watcher between each test. The goal to ensure a clean
state between tests. However, since Watcher can report "yes - I am stopped", but there
are still running Watches, the tests may bleed over into each other, especially on slow
machines.  This can result in errors related to "Watch is already queued in thread pool"
and a state: "not_executed_already_queued", and is VERY difficult to reproduce.

This commit changes the waiting for Watches on stop/pause from an aysnc waiting, back to a
sync wait as it worked prior to #30118.  This help ensure that for testing testing scenario
the stop much more predictable, such that after fully stopped, no Watches are running.
This should have little impact if any on production code since Watcher isn't stopped/paused
too often and when it stop/pause it has the same behavior is the same, it will just run on
the calling thread, not a generic thread.

jakelandis committed:
Merge branch 'watcher_stop_less_async' of github.com:jakelandis/elasticsearch into watcher_stop_less_async

jakelandis commented:
test failures appear relevant... looking into it.

jakelandis commented:
It appears that this can result in holding up the cluster state applier thread too long.  closing this PR and will open a new one that will take into account the concurrent executions in addition to the closed state returned by watcher stats to block until Watcher is fully stopped.
